tower-retry/src/lib.rs

```diff
@@ -37,15 +37,15 @@ impl<P, S> Retry<P, S> {
 
 impl<P, S, Request> Service<Request> for Retry<P, S>
 where
-    P: Policy<Request, S::Response, S::Error> + Clone,
+    P: Policy<Request, S::Response, S::Error> + Clone + Unpin,
```

Comment:
This bound is needed because `Service` implicitly requires `Self: Unpin`, due to the `poll_ready`/`call` conflict.

Reply:
Don't you want just the Future of this to be `Unpin`?

Reply:
It was basically because `ResponseFuture::poll` (which only has `Pin<&mut Self>`) needed to call `Retry::poll_ready`, which takes `&mut self`. That only works (without `unsafe`) if we make `Retry: Unpin`. I worked around it by assuming that `Retry::poll_ready` is really just calling `poll_ready` on the service, and thus all it needs is to get `&mut S`, which it can do since `S` is never pinned. It ain't pretty, but it works.